ostd/src/mm/page_table/cursor.rs

```diff
@@ -526,24 +526,34 @@ where
         }
     }
 
-    /// Unmaps the range starting from the current address with the given length of virtual address.
+    /// Find and remove the first page in the cursor's following range.
     ///
-    /// Already-absent mappings encountered by the cursor will be skipped. It is valid to unmap a
-    /// range that is not mapped.
+    /// The range to be found in is the current virtual address with the
+    /// provided length.
+    ///
+    /// The function stops and yields the page if it has actually removed a
+    /// page, no matter if the following pages are also required to be unmapped.
+    /// The returned page is the virtual page that existed before the removal
+    /// but having just been unmapped.
+    ///
+    /// It also makes the cursor moves forward to the next page after the
+    /// removed one, when an actual page is removed. If no mapped pages exist
+    /// in the following range, the cursor will stop at the end of the range.
     ///
     /// # Safety
     ///
-    /// The caller should ensure that the range being unmapped does not affect kernel's memory safety.
+    /// The caller should ensure that the range being unmapped does not affect
+    /// kernel's memory safety.
     ///
     /// # Panics
     ///
-    /// This function will panic if:
-    ///  - the range to be unmapped is out of the range where the cursor is required to operate;
-    ///  - the range covers only a part of a page.
-    pub unsafe fn unmap(&mut self, len: usize) {
-        let end = self.0.va + len;
+    /// This function will panic if the end range covers a part of a huge page
+    /// and the next page is that huge page.
+    pub unsafe fn take_next(&mut self, len: usize) -> PageTableItem {
```

Comment:
I'm satisfied with the effects of this method, but I wonder if the `len` parameter is necessary? Do we have a scenario where we need to take a cursor of a range but only allow `take_next` (or `unmap`) on one of its subranges? I feel like the semantics of this `len` parameter might make the API look a bit strange, and we also need an additional assert (although minor).

Reply:
The semantic of the `vm_space::CursorMut::unmap(len)` is preserved in this PR. An alternative is to let the `take_next` function search in the rest of the space of the entire cursor, rather than allowing the caller to specify a length. But then we would not be able to implement the `vm_space::CursorMut::unmap(len)` API, unless we change `vm_space::CursorMut::unmap` to `vm_space::CursorMut::take_next`, which is a little bit weird. Now the page table behaves like a collection, but the `VmSpace` is actually a memory space that have mappings to map an un-map, which looks good.